csharp/src/Microsoft.ML.OnnxRuntime/InferenceSession.cs

```diff
@@ -901,8 +901,8 @@ internal IntPtr Handle
 
```

Comment:
As this has unmanaged resources doesn't it need a finalizer?

Reply:
I think it does not. In fact. I suspect the finalizer that I removed contributes to their problems. As they do not dispose properly, finalizer invokes Dispose(bool) potentially when it is too late.

---
In reply to: [476740117](https://github.com/microsoft/onnxruntime/pull/4915#discussion_r476740117) [](ancestors = 476740117)

Reply:
Finalizer in C# is a different concept to a destructor in C++. It's for cleanup of unmanaged resources in case Dispose(true) wasn't called earlier, and may run post GC. 

Ideally Dispose(true) takes care of freeing the resource, but worst case the finalizer can ensure that happens via calling Dispose(false).

---
In reply to: [476749626](https://github.com/microsoft/onnxruntime/pull/4915#discussion_r476749626) [](ancestors = 476749626,476740117)

Reply:
That almost never happen. By the time Finalizer() runs we are out of memory, out of handles and deadlocked if locks are involved. For the finalizer to run, it requires GC to run first, and that required managed memorypressure and that has nothing to do with our native objects,


---
In reply to: [476752910](https://github.com/microsoft/onnxruntime/pull/4915#discussion_r476752910) [](ancestors = 476752910,476749626,476740117)

Reply:
My understanding was that a finalizer being called was a far more common thing and does not only happen in 'we're blowing up' scenarios. 

Given [https://docs.microsoft.com/en-us/dotnet/standard/garbage-collection/unmanaged](https://docs.microsoft.com/en-us/dotnet/standard/garbage-collection/unmanaged) says you should implement a finalizer to ensure unmanaged resources are cleaned up if the user forgets to call Dispose, and all the finalizer needs to do is call Dispose(false), and that prevents any memory leaks (and potential github issues complaining we leak memory even though it's the user's fault for not calling Dispose), what is the benefit to not having them given the cost to implement them seems trivial?

---
In reply to: [476815607](https://github.com/microsoft/onnxruntime/pull/4915#discussion_r476815607) [](ancestors = 476815607,476752910,476749626,476740117)

Reply:
GC frequently gets invoked on 0 generations but objects that have finalizers are promoted to higher generations both for that reason and for the reason that the object may be a long lived object. However, that is not the end of the story,. Even when GC finally gets invoked on such an object the finalizer is not invoked then. It puts it in the finalizer Q. And then when its turn comes along, that finalizer is invoked. The chances are pretty slim.

---
In reply to: [476854054](https://github.com/microsoft/onnxruntime/pull/4915#discussion_r476854054) [](ancestors = 476854054,476815607,476752910,476749626,476740117)

Reply:
[https://docs.microsoft.com/en-us/dotnet/api/system.object.finalize?view=netcore-3](https://docs.microsoft.com/en-us/dotnet/api/system.object.finalize?view=netcore-3).
What is particularly troubling is that objects that depend on each other an even have references to each other, are still finalized in undetermined order. For example, OrtMemoyrAllocation has reference to OrtAllocator but it can get finalized before the allocation. I am not even mentioning if the underlying C++ objects have dependancies on each other.

---
In reply to: [476854054](https://github.com/microsoft/onnxruntime/pull/4915#discussion_r476854054) [](ancestors = 476854054,476815607,476752910,476749626,476740117)

Reply:
Finalizers run late. You're strongly discouraged from ever using them **unless** you have unmanaged memory that isn't using SafeHandle.